File: rb/test/custom_entity_test.rb
# Custom entity test

require "minitest/autorun"
require "json"
require_relative "../HostedRest_sdk"
require_relative "runner"

class CustomEntityTest < Minitest::Test
  def test_create_instance
    testsdk = HostedRestSDK.test(nil, nil)
    ent = testsdk.Custom(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = custom_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "custom." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set HOSTED_REST_TEST_CUSTOM_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    custom_ref01_ent = client.Custom(nil)
    custom_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.custom"), "custom_ref01"))
    custom_ref01_data["path"] = setup[:idmap]["path01"]

    custom_ref01_data_result = custom_ref01_ent.create(custom_ref01_data, nil)
    custom_ref01_data = Helpers.to_map(custom_ref01_data_result.respond_to?(:data_get) ? custom_ref01_data_result.data_get : custom_ref01_data_result)
    assert !custom_ref01_data.nil?

    # UPDATE
    custom_ref01_data_up0_up = {
    }

    custom_ref01_resdata_up0_result = custom_ref01_ent.update(custom_ref01_data_up0_up, nil)
    custom_ref01_resdata_up0 = Helpers.to_map(custom_ref01_resdata_up0_result.respond_to?(:data_get) ? custom_ref01_resdata_up0_result.data_get : custom_ref01_resdata_up0_result)
    assert !custom_ref01_resdata_up0.nil?

    # LOAD
    custom_ref01_match_dt0 = {}
    custom_ref01_data_dt0_loaded = custom_ref01_ent.load(custom_ref01_match_dt0, nil)
    assert !custom_ref01_data_dt0_loaded.nil?


  end
end

def custom_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "custom", "CustomTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = HostedRestSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["custom01", "custom02", "custom03", "path01"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["HOSTED_REST_TEST_CUSTOM_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "HOSTED_REST_TEST_CUSTOM_ENTID" => idmap,
    "HOSTED_REST_TEST_LIVE" => "FALSE",
    "HOSTED_REST_TEST_EXPLAIN" => "FALSE",
    "HOSTED_REST_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["HOSTED_REST_TEST_CUSTOM_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["HOSTED_REST_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["HOSTED_REST_APIKEY"],
      },
      extra || {},
    ])
    client = HostedRestSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["HOSTED_REST_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["HOSTED_REST_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/custom_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/hosted-rest-sdk/go"
	"github.com/voxgig-sdk/hosted-rest-sdk/go/core"

	vs "github.com/voxgig-sdk/hosted-rest-sdk/go/utility/struct"
)

func TestCustomEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Custom(nil)
		if ent == nil {
			t.Fatal("expected non-nil CustomEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := customBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "custom." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set HOSTED_REST_TEST_CUSTOM_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		customRef01Ent := client.Custom(nil)
		customRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "custom"}, setup.data), "custom_ref01"))
		customRef01Data["path"] = setup.idmap["path01"]

		customRef01DataResult, err := customRef01Ent.Create(customRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		customRef01Data = core.ToMapAny(entityData(customRef01DataResult))
		if customRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if customRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		customRef01DataUp0Up := map[string]any{
			"id": customRef01Data["id"],
		}

		customRef01ResdataUp0Result, err := customRef01Ent.Update(customRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		customRef01ResdataUp0 := core.ToMapAny(entityData(customRef01ResdataUp0Result))
		if customRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if customRef01ResdataUp0["id"] != customRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}

		// LOAD
		customRef01MatchDt0 := map[string]any{
			"id": customRef01Data["id"],
		}
		customRef01DataDt0Loaded, err := customRef01Ent.Load(customRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		customRef01DataDt0LoadResult := core.ToMapAny(entityData(customRef01DataDt0Loaded))
		if customRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if customRef01DataDt0LoadResult["id"] != customRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		customRef01MatchRm0 := map[string]any{
			"id": customRef01Data["id"],
		}
		_, err = customRef01Ent.Remove(customRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func customBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "custom", "CustomTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read custom test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse custom test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"custom01", "custom02", "custom03", "path01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("HOSTED_REST_TEST_CUSTOM_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"HOSTED_REST_TEST_CUSTOM_ENTID": idmap,
		"HOSTED_REST_TEST_LIVE":      "FALSE",
		"HOSTED_REST_TEST_EXPLAIN":   "FALSE",
		"HOSTED_REST_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["HOSTED_REST_TEST_CUSTOM_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["HOSTED_REST_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["HOSTED_REST_APIKEY"],
			},
			extra,
		})
		client = sdk.NewHostedRestSDK(core.ToMapAny(mergedOpts))
	}

	live := env["HOSTED_REST_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["HOSTED_REST_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
